### DISCUSSION TITLE
How/where the decompiler functions

### DISCUSSION THREAD

**Discussion by Wall-AF** [NationalSecurityAgency/ghidra#2899] [Q&A]
Can anyone explain where in the code does Ghidra turn this assembly
![image](https://user-images.githubusercontent.com/67698927/113587617-f481c100-9626-11eb-8787-d98d8c352f2e.png)
into this C
![image](https://user-images.githubusercontent.com/67698927/113586071-f64a8500-9624-11eb-9237-ff93e8eb740c.png)
?

I'm trying to determine where to fix this?  It occurs because I've had to increase the size of the `ptr` that points to the object!

**Comment by hedgeberg** [comment]
There is no single place where the decompiler converts a set of opcodes into a line of code. The whole thing is a slow process involving resolution of variables and trying to recover code snippets. 

The decompiler's core logic "loop" could best be described as a "reverse optimizer", where the deconpiler iterates over 2 structures (the Varnode tree and the Pcodeop flow graph) to try and convert the complex semantics that are present in machine code back into something closer to what the first step of forward compilation would output. In essence, it attempts to take this big, really nasty data structure and trim as much of the fat as it can in the opposite way to how a forward compiler optimizes the assembly that it generates. Ideally, at the end of this, you have what's referred to by compiler & program analysis people as an "intermediate representation": a maximally-simplified representation of the steps the code takes that produce lasting effects on varnodes. Then, that representation is fed into a reverse-parser to generate clang tokens, which are turned into the code you see. 

The issue here isn't because of the decompiler -- either your int, uint, or pointer size are inconsistent,  your variable newCplex is not a UINT* type, or your UINT type isn't 4 bytes wide.

**Reply by Wall-AF** [reply]
@hedgeberg thanks.  I kinda got that far!  Was hoping for some more insight into it.  My issue concerns different sized pointers, and I can see there's been substantial work done on this in the disassembler, it just hasn't been followed through all the way.

**Reply by Wall-AF** [reply]
I've discovered that the expected `CPUI_PTRSUB` op is not being added to the process.

**Reply by Wall-AF** [reply]
Interestingly, if I change the instruction at `1158:0f21` to `MOV  word ptr ES:[BX + `**0x0**`],pNewCplex` it is recognised as being the first element of the structure.  This implies that there's a mistake in one of the rules that add offsets to segmented pointers.  Anyone know which???

**Reply by Wall-AF** [reply]
Removing the `RuleDumptyHump` stops the correct interpretation when using the 16-bit pointer.  Implying that this rule is necessary for correct interpretation, however, it never gets triggered when using the 32-bit pointer.

**Reply by Wall-AF** [reply]
I think the problem is that in the early stages of the `pcode` generation, the `word ptr ES:[BX + 0x2]` needs to be treated as `word ptr [ES:BX + 0x2]` for when we explictly size our parameter pointer to be 32-bit.

Anyone any ideas how to do this, or am I barking up the wrong tree???

**Comment by Wall-AF** [comment]
I think I've solved this one, but I would like someone (in the know) to please take a look.

Within the C++ function `int4 RuleSegment::applyOp(PcodeOp *op,Funcdata &data)`, providing the segment meets the `hasFarPointerSupport()`, it merges 2 varnodes (`vn1`, `vn2`) if and only if both defining `PcodeOp`'s code (`vn?->getDef()->code()`) is `CPUI_SUBPIECE`.  However, when something is added to the 2nd varnode `vn2` (say to reference part of a structure), that `vn2->getDef()->code()` isn't `CPUI_SUBPIECE`.  However its 0'th input's defining `PcodeOp` (`vn2->getDef()-getIn(0)->getDef()->code()`) is.

By changing that rule to allow both options sorts this out.

2 files needed modifying:
```
D:\GIT_Sources\Ghidra\ghidra\Ghidra\Features\Decompiler\src\decompile\cpp>git diff ruleaction.cc
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc
index a1c01498e..e6adbade9 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/ruleaction.cc
@@ -7467,7 +7467,14 @@ int4 RuleSegment::applyOp(PcodeOp *op,Funcdata &data)
     data.opRemoveInput(op,2);
     data.opRemoveInput(op,1);
     data.opSetInput(op,whole,0);
-    data.opSetOpcode(op,CPUI_COPY);
+    //data.opSetOpcode(op, CPUI_COPY);
+    PcodeOp *op2 = vn2->getDef();
+    if (op2->code() != CPUI_SUBPIECE) {
+      data.opInsertInput(op,op2->getIn(1),1);
+      data.opSetOpcode(op,op2->code());
+    } else {
+      data.opSetOpcode(op,CPUI_COPY);
+    }
     return 1;
   }
   return 0;

D:\GIT_Sources\Ghidra\ghidra\Ghidra\Features\Decompiler\src\decompile\cpp>git diff varnode.cc
diff --git a/Ghidra/Features/Decompiler/src/decompile/cpp/varnode.cc b/Ghidra/Features/Decompiler/src/decompile/cpp/varnode.cc
index 919230507..244b8ac51 100644
--- a/Ghidra/Features/Decompiler/src/decompile/cpp/varnode.cc
+++ b/Ghidra/Features/Decompiler/src/decompile/cpp/varnode.cc
@@ -1594,7 +1594,12 @@ bool contiguous_test(Varnode *vn1,Varnode *vn2)
   Varnode *vnwhole;
   switch(op1->code()) {
   case CPUI_SUBPIECE:
-    if (op2->code() != CPUI_SUBPIECE) return false;
+    //if (op2->code() != CPUI_SUBPIECE) return false; // original test
+    if (op2->code() != CPUI_SUBPIECE) {
+      if ((Varnode*)0 == op2->getIn(0)) return false;
+      op2 = op2->getIn(0)->getDef();
+      if (op2->code() != CPUI_SUBPIECE) return false;
+    }
     vnwhole = op1->getIn(0);
     if (op2->getIn(0) != vnwhole) return false;
     if (op2->getIn(1)->getOffset() != 0)
```

**Comment by mattiasgrenfeldt** [comment]
How have you turned on the P-code output underneath the assembly instructions? Also is that P-code after it has been simplified by the decompiler or just after it has been lifted from machine code?

**Reply by Wall-AF** [reply]
To display the PCode:
1. Move the cursor to an asssembly instruction in the `Listing:` panel
2. Press the `Edit the Listing fields` button (see below - the 4th button)
3. Right-Click the `PCode` button
4. Choose `Enable Field`.
![image](https://user-images.githubusercontent.com/67698927/224671432-a2041782-6f92-4b33-973c-f6b5609e4740.png)

Is it the simplified PCode?
- I'm unsure.

**Reply by mattiasgrenfeldt** [reply]
Wow, that's a button I had never noticed, thank you! :)

**Reply by astrelsky** [reply]
> Is it the simplified PCode?
> - I'm unsure.

It is not the simplified pcode. As far as I know, there is no builtin way to see the simplified pcode within the Ghidra UI. I do recall seeing a plugin which provides this recently.
